libcontainer/container_linux.go

```diff
@@ -2826,33 +2827,47 @@ func skipShiftfsBindSource(source string) bool {
 }
 
 // sysbox-runc: determines which mounts must be ID-mapped; does not actually
-// perform the ID-mapped mounts, simply adds the list of ID-mapped mounts to the
-// config.
+// perform the ID-mapped mounts (that's done inside the container, see
+// rootfs_init_linux.go) but rather marks the mount for ID-mapping only.
 func (c *linuxContainer) setupIDMappedMounts() error {
 
 	config := c.config
 
-	// Determine if other host directories mounted into the container's rootfs
-	// need ID-mapping.
+	// rootfs
+	if config.RootfsUidShiftType == sh.IDMappedMount ||
+		config.RootfsUidShiftType == sh.IDMappedMountOrShiftfs {
+		idMapMountAllowed, err := sh.IDMapMountSupportedOnPath(config.Rootfs)
+		if err != nil {
+			return newSystemErrorWithCausef(err, "checking for ID-mapped mount support on rootfs %s", config.Rootfs)
+		}
+		if idMapMountAllowed {
+			config.RootfsUidShiftType = sh.IDMappedMount
+		}
+	}
+
+	// bind-mounts
+	if config.BindMntUidShiftType == sh.IDMappedMount ||
+		config.BindMntUidShiftType == sh.IDMappedMountOrShiftfs {
 
-	for _, m := range config.Mounts {
-		if m.Device == "bind" {
+		for _, m := range config.Mounts {
+			if m.Device == "bind" {
 
-			idMapMntAllowed, err := sh.IDMapMountSupportedOnPath(m.Source)
-			if err != nil {
-				return newSystemErrorWithCausef(err, "checking for ID-mapped mount support on bind source %s", m.Source)
-			}
+				idMapMntAllowed, err := sh.IDMapMountSupportedOnPath(m.Source)
+				if err != nil {
+					return newSystemErrorWithCausef(err, "checking for ID-mapped mount support on bind source %s", m.Source)
+				}
 
-			if !idMapMntAllowed {
-				continue
-			}
+				if !idMapMntAllowed {
```

Comment:
Sorry for the confusion; in this function we are only looking to set up ID-mapped mounts (not shiftfs). Therefore, the code says: if ID-mapped mount is not supported on a given container mount, nothing else to do for that mount (i.e., `continue`). Otherwise, the code says: use ID-mapping if the mount really needs "uid shifting". By "uid shifting" I don't mean shiftfs specifically, but rather "the mount is not in the container's uid:gid range and therefore needs some way to shift ownership (in this case is via ID-mapping)".

I think the word "shift" in `needUidShiftOnBindSrc()` is confusing you because you are relating it to shiftfs (?)

Reply:
Oh, I see, that makes sense, as you said I was confused by the "shift" keyword in `needUidShiftOnBindSrc()` routine. There's no need for you to do anything here, I get it now.